File: src/main/java/test/DataProcessingInMemory.java

package test;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

public class DataProcessingInMemory implements IDataProcessing {

    public final ConcurrentMap<String, Integer> SCORE_MAP = new ConcurrentHashMap<>();
    public final ConcurrentMap<String, List> PHRASE_MAP = new ConcurrentHashMap<>();

    @Override
    public boolean savePalindrome(String userName, String phrase) {
        synchronized (PHRASE_MAP) {
            List<String> phraseList = PHRASE_MAP.getOrDefault(userName, new ArrayList());
            PHRASE_MAP.putIfAbsent(userName, phraseList);

            if (phraseList.contains(phrase)) {
                return false;
            } else {
                phraseList.add(phrase);
                return true;
            }
        }
    }

    @Override
    public int addScore(String userName, int score) {
        return SCORE_MAP.compute(userName, (k, v) -> v == null ? score : v + score);
    }
    
    @Override
    public int getScore(String userName) {
        return SCORE_MAP.getOrDefault(userName, 0);
    }
    
    @Override
    public Map getTopUsers() {
        Map<String, Integer> map = new HashMap<>();
        synchronized (SCORE_MAP) {
            SCORE_MAP.entrySet().stream()
                    .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
                    .limit(5)
                    .forEach(e -> map.put(e.getKey(), e.getValue()));
        }
        return map;
    }
    
}

File: src/main/scala/test/DataProcessingInMemory.scala
package test;

import java.util.{ArrayList, List, Comparator}
import java.util.concurrent.ConcurrentHashMap
import collection.JavaConverters._

class DataProcessingInMemory extends DataProcessing {

  private val SCORE_MAP = new ConcurrentHashMap[String, Int].asScala
  private val PHRASE_MAP = new ConcurrentHashMap[String, Seq[String]].asScala

  override def savePalindrome(userName: String, phrase: String): Either[OldScore, NewScore] = 
    PHRASE_MAP.synchronized {
      val phraseList = PHRASE_MAP.getOrElseUpdate(userName, Seq[String]())

      if (phraseList.contains(phrase)) 
        Left(sumSeqElemSize(phraseList))
      else {
        val newPhraseList = phraseList :+ phrase
        PHRASE_MAP.put(userName, newPhraseList)        
        Right(sumSeqElemSize(newPhraseList))
      }
    }    
  
  override def getScore(userName: String): Int = 
    PHRASE_MAP.get(userName) match {
      case Some(seq) => sumSeqElemSize(seq)
      case _ => 0
  }
   
  override def getUsersAndScores = {
    val res = PHRASE_MAP.map{case (u, l) => (u, sumSeqElemSize(l))}
    collection.immutable.Map(res.toSeq: _*)
  }
  
  private def sumSeqElemSize(seq: Seq[String]) = seq.foldLeft(0)((a, s) => a + s.size)  

}
